python/cuvs/cuvs/neighbors/hnsw/hnsw.pyx

```diff
@@ -0,0 +1,375 @@
+    saved as a built CAGRA index. The loaded index is immutable and can only
+    be searched by the hnswlib wrapper in cuVS, as the format is not
+    compatible with the original hnswlib.
+
+    Saving / loading the index is experimental. The serialization format is
+    subject to change, therefore loading an index saved with a previous
+    version of cuVS is not guaranteed to work.
+
+    Parameters
+    ----------
+    filename : string
+        Name of the file.
+    dim : int
+        Dimensions of the training dataest
+    dtype : np.dtype of the saved index
+        Valid values for dtype: [np.float32, np.byte, np.ubyte]
+    metric : string denoting the metric type, default="sqeuclidean"
+        Valid values for metric: ["sqeuclidean", "inner_product"], where
+            - sqeuclidean is the euclidean distance without the square root
+              operation, i.e.: distance(a,b) = \\sum_i (a_i - b_i)^2,
+            - inner_product distance is defined as
+              distance(a, b) = \\sum_i a_i * b_i.
+    {resources_docstring}
+
+    Returns
+    -------
+    index : HnswIndex
+
+    Examples
+    --------
+    >>> import cupy as cp
+    >>> from cuvs.neighbors import cagra
+    >>> from cuvs.neighbors import hnsw
+    >>> n_samples = 50000
+    >>> n_features = 50
+    >>> dataset = cp.random.random_sample((n_samples, n_features),
+    ...                                   dtype=cp.float32)
+    >>> # Build index
+    >>> index = cagra.build(cagra.IndexParams(), dataset)
+    >>> # Serialize the CAGRA index to hnswlib base layer only index format
+    >>> hnsw.save("my_index.bin", index)
+    >>> index = hnsw.load("my_index.bin", n_features, np.float32,
+    ...                   "sqeuclidean")
+    """
+    cdef Index idx = Index()
+    cdef cuvsResources_t res = <cuvsResources_t>resources.get_c_obj()
+    cdef string c_filename = filename.encode('utf-8')
+    cdef cydlpack.DLDataType dl_dtype
+    if dtype == np.float32:
+        dl_dtype.code = cydlpack.kDLFloat
+        dl_dtype.bits = 32
+        dl_dtype.lanes = 1
+    elif dtype == np.ubyte:
+        dl_dtype.code = cydlpack.kDLUInt
+        dl_dtype.bits = 8
+        dl_dtype.lanes = 1
+    elif dtype == np.byte:
+        dl_dtype.code = cydlpack.kDLInt
+        dl_dtype.bits = 8
+        dl_dtype.lanes = 1
+    else:
+        raise ValueError("Only float32 is supported for dtype")
+
+    idx.index.dtype = dl_dtype
+    cdef cuvsDistanceType distance_type = DISTANCE_TYPES[metric]
+
+    check_cuvs(cuvsHnswDeserialize(
+        res,
+        c_filename.c_str(),
+        dim,
+        distance_type,
+        idx.index
+    ))
+    idx.trained = True
+    return idx
+
+
+@auto_sync_resources
+def from_cagra(cagra.Index index, resources=None):
+    """
+    Returns an hnsw base-layer-only index from a CAGRA index.
+
+    NOTE: This method uses the filesystem to write the CAGRA index in
+          `/tmp/<random_number>.bin` before reading it as an hnsw index,
```

Comment:
Can we please make this path configurable? It's not safe to assume we always have access to this directory (not that this directory is always where the user wants to store this intermediate file). We should provide an argument to control this path but allow the user to override this path if they need/want to.